### PR TITLE
JobId-based remote writeback

### PR DESCRIPTION
**Pull request by DarthMax**
- **Introduce GdsArrowClient**
- **Use GdsArrowClient in ArrowGraphConstruction and ArrowQueryRunner**
- **Adapt client to push based remote projection**
- **Fix most style and typing issues**

### CODE REVIEW

**Comment by DarthMax** [comment]
[[🍓] Adjust python client to the new arrow architecture](https://trello.com/c/oAPuxuI9/9292-%F0%9F%8D%93-adjust-python-client-to-the-new-arrow-architecture)

File: graphdatascience/query_runner/aura_db_arrow_query_runner.py [hunk 1]
```diff
@@ -128,57 +96,133 @@ def create_graph_constructor(
...
+
+    @staticmethod
+    def _extract_write_back_arguments(proc_name: str, params: dict[str, Any]) -> dict[str, Any]:
+        config = params.get("config", {})
+        write_config = {}
+
+        if "concurrency" in config:
+            write_config["concurrency"] = config["concurrency"]
+
+        if "gds.shortestPath" in proc_name:
```
Comment (soerenreichardt) [inline graphdatascience/query_runner/aura_db_arrow_query_runner.py:182]: we should also check for `gds.allShortestPath`

File: graphdatascience/query_runner/aura_db_arrow_query_runner.py [hunk 2]
```diff
@@ -1,55 +1,33 @@
...
 
 from ..call_parameters import CallParameters
 from ..session.dbms_connection_info import DbmsConnectionInfo
+from .gds_arrow_client import GdsArrowClient
 from .query_runner import QueryRunner
 from graphdatascience.query_runner.graph_constructor import GraphConstructor
 from graphdatascience.server_version.server_version import ServerVersion
 
 
 class AuraDbArrowQueryRunner(QueryRunner):
```
Comment (FlorentinD) [inline graphdatascience/query_runner/aura_db_arrow_query_runner.py:14]: (side comment)
We should think of changing the class name.

Its not really against arrow endpoints on the DB anymore.

**Review by FlorentinD — COMMENTED** [review]
Nice refactoring of the flight client logic.

Mostly minor comments around the write-back logic.

File: graphdatascience/query_runner/aura_db_arrow_query_runner.py [hunk 3]
```diff
@@ -128,57 +96,133 @@ def create_graph_constructor(
...
+
+            elif "gds.graph.nodeLabel.write" == proc_name:
+                write_config["nodeLabels"] = [params["node_label"]]
+
+            elif "gds.graph.relationshipProperties.write" == proc_name:
+                write_config["relationshipProperties"] = params["relationship_properties"]
+                write_config["relationshipType"] = params["relationship_type"]
+
+            elif "gds.graph.relationship.write" == proc_name:
+                if "relationship_property" in params and params["relationship_property"] != "":
```
[the remark this answers is not in the document]
Comment (FlorentinD) [reply graphdatascience/query_runner/aura_db_arrow_query_runner.py:208]: ah thats just the positional parameter in the string

File: graphdatascience/graph/base_graph_proc_runner.py [hunk 4]
```diff
@@ -517,7 +517,8 @@ def writeRelationship(
         ).squeeze()
 
     @multimethod
-    def removeNodeProperties(self) -> None: ...
+    def removeNodeProperties(self) -> None:
+        ...
```
[the remark this answers is not in the document]
Comment (DarthMax) [reply graphdatascience/graph/base_graph_proc_runner.py:521]: I updated an ran again. Seems to have change this back :+1:

File: graphdatascience/query_runner/aura_db_arrow_query_runner.py [hunk 5]
```diff
@@ -128,57 +96,133 @@ def create_graph_constructor(
...
+        params: CallParameters,
+        yields: Optional[List[str]] = None,
+        database: Optional[str] = None,
+        logging: bool = False,
+        custom_error: bool = True,
+    ) -> DataFrame:
+        host, port = self._gds_arrow_client.connection_info()
+        token = self._gds_arrow_client.get_or_request_token()
+        if token is None:
+            token = "IGNORED"
```
[the remark this answers is not in the document]
Comment (DarthMax) [reply graphdatascience/query_runner/aura_db_arrow_query_runner.py:115]: Currently the token is a required parameter. We could change it but on the other hand it does not really hurt

File: graphdatascience/query_runner/aura_db_arrow_query_runner.py [hunk 6]
```diff
@@ -128,57 +96,133 @@ def create_graph_constructor(
...
+
+        return gds_write_result
+
+    @staticmethod
+    def _extract_write_back_arguments(proc_name: str, params: dict[str, Any]) -> dict[str, Any]:
+        config = params.get("config", {})
+        write_config = {}
+
+        if "concurrency" in config:
+            write_config["concurrency"] = config["concurrency"]
```
[the remark this answers is not in the document]
Comment (DarthMax) [reply graphdatascience/query_runner/aura_db_arrow_query_runner.py:180]: I think I would actually have to check for both :+1:

File: graphdatascience/query_runner/aura_db_arrow_query_runner.py [hunk 7]
```diff
@@ -128,57 +96,133 @@ def create_graph_constructor(
...
+                write_config["relationshipType"] = config["writeRelationshipType"]
+                if "writeProperty" in config:
+                    write_config["relationshipProperties"] = [config["writeProperty"]]
+            else:
+                if "writeProperty" in config:
+                    write_config["nodeProperties"] = [config["writeProperty"]]
+                if "nodeLabels" in params:
+                    write_config["nodeLabels"] = params["nodeLabels"]
+                else:
+                    write_config["nodeLabels"] = ["*"]
```
[the remark this answers is not in the document]
Comment (DarthMax) [reply graphdatascience/query_runner/aura_db_arrow_query_runner.py:226]: It is not the default currently there is a bug in the Neo4j procedure

**Comment by Mats-SX** [comment]
[[🍓] JobId-based remote writeback in client](https://trello.com/c/491gaKQh/9511-%F0%9F%8D%93-jobid-based-remote-writeback-in-client)

**Comment by soerenreichardt** [comment]
Blocked waiting for the 5.21 release

**Pull request merged**